grafana/resource_folder_permission.go

```diff
@@ -0,0 +1,175 @@
+							Optional: true,
+							Default:  -1,
+						},
+						"permission": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringInSlice([]string{"View", "Edit", "Admin"}, false),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func UpdateFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+
+	v, ok := d.GetOk("permissions")
+	if !ok {
+		return nil
+	}
+	permissionList := gapi.PermissionItems{}
+	for _, permission := range v.(*schema.Set).List() {
+		permission := permission.(map[string]interface{})
+		permissionItem := gapi.PermissionItem{}
+		if permission["role"].(string) != "" {
+			permissionItem.Role = permission["role"].(string)
+		}
+		if permission["team_id"].(int) != -1 {
+			permissionItem.TeamId = int64(permission["team_id"].(int))
+		}
+		if permission["user_id"].(int) != -1 {
+			permissionItem.UserId = int64(permission["user_id"].(int))
+		}
+		permissionItem.Permission = mapPermissionStringToInt64(permission["permission"].(string))
+		permissionList.Items = append(permissionList.Items, &permissionItem)
+	}
+
+	folderUID := d.Get("folder_uid").(string)
+
+	err := client.UpdateFolderPermissions(folderUID, &permissionList)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(folderUID)
+
+	return ReadFolderPermissions(d, meta)
+}
+
+func ReadFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gapi.Client)
+
+	folderUID := d.Get("folder_uid").(string)
+
+	folderPermissions, err := client.FolderPermissions(folderUID)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "status: 404") {
+			log.Printf("[WARN] removing folder %s from state because it no longer exists in grafana", folderUID)
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	permissionItems := make([]interface{}, len(folderPermissions))
+	count := 0
+	for _, permission := range folderPermissions {
+		if permission.FolderUid != "" {
+			permissionItem := make(map[string]interface{})
+			permissionItem["role"] = permission.Role
+			permissionItem["team_id"] = permission.TeamId
+			permissionItem["user_id"] = permission.UserId
+			permissionItem["permission"] = mapPermissionInt64ToString(permission.Permission)
+
+			permissionItems[count] = permission
+			count++
+		}
+	}
+
+	d.Set("permissions", permissionItems)
+
+	return nil
+}
+
+func DeleteFolderPermissions(d *schema.ResourceData, meta interface{}) error {
+	//since permissions are tied to folders, we can't really delete the permissions.
+	//we will simply remove all permissions, leaving a folder that only an admin can access.
+	//if for some reason the parent folder doesn't exist, we'll just ignore the error
+	client := meta.(*gapi.Client)
+
+	folderUID := d.Get("folder_uid").(string)
+	emptyPermissions := gapi.PermissionItems{}
+
+	err := client.UpdateFolderPermissions(folderUID, &emptyPermissions)
+	if err != nil {
+		if err.Error() == "404 Not Found" {
```

Comment:
```suggestion
		if strings.HasPrefix(err.Error(), "status: 404") {
```

Reply:
Doh! Good catch... I keep copying that from other places. Went ahead and updated the resource_folder.go since that's where I copied it from.